packages/vite/src/node/server/sourcemap.ts

```diff
@@ -19,39 +19,49 @@ interface SourceMapLike {
   sourceRoot?: string
 }
 
-export async function injectSourcesContent(
-  map: SourceMapLike,
-  file: string,
-  logger: Logger,
-): Promise<void> {
+async function computeSourceRoute(map: SourceMapLike, file: string) {
   let sourceRoot: string | undefined
   try {
     // The source root is undefined for virtual modules and permission errors.
     sourceRoot = await fsp.realpath(
       path.resolve(path.dirname(file), map.sourceRoot || ''),
     )
   } catch {}
+  return sourceRoot
+}
+
+export async function injectSourcesContent(
+  map: SourceMapLike,
+  file: string,
+  logger: Logger,
+): Promise<void> {
+  let sourceRootPromise: Promise<string | undefined>
 
   const missingSources: string[] = []
   const sourcesContent = map.sourcesContent || []
   await Promise.all(
-    map.sources.map(async (sourcePath, index) => {
-      let content = null
-      if (sourcePath && !virtualSourceRE.test(sourcePath)) {
+    map.sources
+      .filter(
+        (sourcePath, index) =>
+          !sourcesContent[index] &&
+          sourcePath &&
+          !virtualSourceRE.test(sourcePath),
+      )
+      .map(async (sourcePath, index) => {
+        // inject content from source file when sourcesContent is null
+        sourceRootPromise ??= computeSourceRoute(map, file)
+        const sourceRoot = await sourceRootPromise
         sourcePath = decodeURI(sourcePath)
         if (sourceRoot) {
           sourcePath = path.resolve(sourceRoot, sourcePath)
         }
-        // inject content from source file when sourcesContent is null
-        content =
-          sourcesContent[index] ??
-          (await fsp.readFile(sourcePath, 'utf-8').catch(() => {
+        sourcesContent[index] = await fsp
```

Comment:
I think this `index` is not correct in some cases because `.filter` changes the index.

```js
[0,1,2,3].filter(v => v !== 0).map((v, i) => console.log(v, i))
// 1 0
// 2 1
// 3 2
```

Reply:
Oh, good catch! Sorry about it, it should be fixed at https://github.com/vitejs/vite/pull/15207/commits/1b2d1e0dbcdbbe29aaac898d001591a128894747